### PR TITLE
Added UnitTests for WorldAnchor.

### PR DESCRIPTION
**Pull request by eluchsinger**
I have added the promised UnitTests.

(#1074)

### CODE REVIEW

**Comment by eluchsinger** [comment]
I was not able to reproduce that error. What could be the cause my tests run through? It is hard for me to know if I am doing it right like this.
I have added a Setup method to clear the scene and add a `WorldAnchorManager` to the scene before the tests run. Can you tell me if it fixed your problems?

**Comment by DDReaper** [comment]
They try the test on a clean scene.  If the test requires a game object setup, then the test initialisation should add it to the scene.   The other tests seem to do this as I tested with a blank scene.
Hope that helps.

File: Assets/HoloToolkit-UnitTests/Editor/Utilities/Extensions/WorldAnchorManagerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace HoloToolkit.Unity.Tests.Extensions
+{
+    /// <summary>
+    /// This class extends the world anchor manager with methods constrained to test logic.
+    /// </summary>
+    public static class WorldAnchorManagerExtensions
+    {
+        internal static string GenerateAnchorName(this WorldAnchorManager anchorManager, GameObject gameObjectToAnchor, string anchorName = null)
+        {
+            return string.IsNullOrEmpty(anchorName) ? gameObjectToAnchor.name : anchorName;
```
Comment (Zod-) [inline Assets/HoloToolkit-UnitTests/Editor/Utilities/Extensions/WorldAnchorManagerExtensions.cs]: Did you only copy this logic and leave it in the anchor manager? If so, this method is only ever called in the tests, which doesn't really make sense. 
This definetly has to be in the main folder, not in `UnitTests`. There also is no need for this to be an extension, as the anchormanager isn't used. I would either put it in `WorldAnchorManager` directly or make a `WorldAnchorUtilities` and put it in there where both version would be `public static`. Like that you wouldn't need to initialise the `WorldAnchorManager` for the test as well.

In conclusion:
1. Move this logic out of the `AttachAnchor` Method
2. Make a public static Method in either WorldAnchorManager or WorldAnchorUtilities
3. Test that instead, which should make the setup a lot easier too.

File: Assets/HoloToolkit-UnitTests/Editor/Utilities/WorldAnchorManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+namespace HoloToolkit.Unity.Tests
+{
+    public class WorldAnchorManagerTests
+    {
+        [SetUp]
+        public void ClearScene()
+        {
+            TestUtils.ClearScene();
+            var managers = new GameObject("Managers");
+            managers.AddComponent<WorldAnchorManager>();
```
Comment (Zod-) [inline Assets/HoloToolkit-UnitTests/Editor/Utilities/WorldAnchorManagerTests.cs]: This won't be needed with the other changes above however, you could make this work with:
```chsarp
managers.CallAwake().CallStart();
```
Explanation [here](https://github.com/Microsoft/MixedRealityToolkit-Unity/tree/master/Assets/HoloToolkit-UnitTests#2-any-monobehaviour-methods-such-as-awake-start-and-update-will-not-be-called-by-unity-in-tests)

Comment (eluchsinger) [reply Assets/HoloToolkit-UnitTests/Editor/Utilities/WorldAnchorManagerTests.cs]: I read the documentation, but I didn't understand when to use these methods. Thanks for the example.

File: Assets/HoloToolkit-UnitTests/Editor/Utilities/Extensions/WorldAnchorManagerExtensions.cs (hunk 1 above)
Comment (eluchsinger) [reply Assets/HoloToolkit-UnitTests/Editor/Utilities/Extensions/WorldAnchorManagerExtensions.cs]: The logic is duplicated code. It is both in the test method and in the `WorldAnchorManager`, which is kind of bad. Do you think I should create a `WorldAnchorUtilities` in the main folder? This class would have only one method as of now... Probably it is better, if I take the method into the `WorldAnchorManager` directly and make it `public static`. I think, I was just afraid of changing too much 👎 .

**Comment by Zod-** [comment]
@eluchsinger I think it's fine like this. I'm undecided whether to move the logic into a separate utility class or not. Maybe someone else has some input on that.

**Comment by DDReaper** [comment]
Ideally we should not have duplicated cod for any reason.
If the code is used by both test and apps, it should be in or as an extension of the main class.
If it is only used by tests, then it should only exist in the tests project so that it does not contaminate the main code stream.

So does this extension function have any use outside of tests?

**Comment by DDReaper** [comment]
I'll give this another test now if you think it's ready @eluchsinger ?

**Comment by eluchsinger** [comment]
@DDReaper yes, I thought it had to be duplicated, because that was the only way of testing it without initializing the `WorldAnchorManager`. Zod said in #1074 that I should extract the method's logic because it won't run in the editor, and I somehow misinterpreted it as to put it directly into the tests. I fixed that now and it shouldn't need to initialize a WorldAnchorManager object.

I couldn't reproduce your test failure, so I am not 100% sure that this time the test won't fail, but I believe the code is now at the right place. Please re-run the tests, I think it should work now.

**Comment by Zod-** [comment]
@eluchsinger `AttachAnchor` will just return null in the editor. That's what I meant with that it won't run in the editor. You had to extract the logic because it was in the `#else` block of this compiler directive.

https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/bf98b15bd7ce6154e9366bf5368a2bf09888dbd8/Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs#L183-L188

**Comment by DDReaper** [comment]
Was going to ask for another test to ensure thenegative case would be true but then noticed a possible error.  If you call this function with a new proposed name, it is NOT stored against the anchor, only returned.
Is that the intended behaviour
```
        public static string GenerateAnchorName(GameObject gameObjectToAnchor, string proposedAnchorname = null)
        {
            return string.IsNullOrEmpty(proposedAnchorname) ? gameObjectToAnchor.name : proposedAnchorname;
        }
```
so calling this function currently : 
```
var name = GenerateAnchorName(anchor,"newname")
```
is exactly the same as: 
```
var name = newname
```

It doesn't update the name of the anchor itself??

**Comment by DDReaper** [comment]
Ignore my last comment.  I'm now ok with this change.

File: Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs [hunk 3]
```diff
@@ -170,6 +170,17 @@ private void Anchor_OnTrackingChanged(WorldAnchor anchor, bool located)
 
         #endregion // Event Callbacks
 #endif
+        /// <summary>
+        /// Generates the name for the anchor.
+        /// If no anchor name was specified, the name of the anchor will be the same as the GameObject's name.
+        /// </summary>
+        /// <param name="gameObjectToAnchor">The GameObject to attach the anchor to.</param>
+        /// <param name="proposedAnchorname">Name of the anchor. If none provided, the name of the GameObject will be used.</param>
+        /// <returns>The name of the newly attached anchor.</returns>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs]: The anchor is never attached in this method, so I don't think this comment is quite right,
